module: print amount of load time of a module

H4ad opened the pull request:
I was investigating why `npm` is slow and I started to figure out that was caused by it loading a lot of modules/js files, but I didn't know what are those modules or even how much time it took to load them.

So, this implementation came to my mind, and I used with the following commands:

```bash
NODE_DEBUG=module_cjs ~/Projects/opensource/node-copy-4/out/Release/node ./bin/npm-cli.js run echo --silent &> trace-node.txt
```

The log looks like this:

```
MODULE_CJS 413480 [/home/h4ad/Projects/opensource/performance-analysis/npm-cli/lib/commands/run-script.js] [../utils/did-you-mean.js]: 0.499ms
MODULE_CJS 413480 [/home/h4ad/Projects/opensource/performance-analysis/npm-cli/lib/commands/run-script.js] [../utils/is-windows.js]: 0.102ms
MODULE_CJS 413480 [/home/h4ad/Projects/opensource/performance-analysis/npm-cli/lib/base-command.js] [path]: 0.005ms
MODULE_CJS 413480 [/home/h4ad/Projects/opensource/performance-analysis/npm-cli/lib/base-command.js] [@npmcli/config/lib/definitions]: 0.001ms
MODULE_CJS 413480 [/home/h4ad/Projects/opensource/performance-analysis/npm-cli/lib/workspaces/get-workspaces.js] [path]: 0.003ms
MODULE_CJS 413480 [/home/h4ad/Projects/opensource/performance-analysis/npm-cli/lib/workspaces/get-workspaces.js] [@npmcli/map-workspaces]: 0.051ms
MODULE_CJS 413480 [/home/h4ad/Projects/opensource/performance-analysis/npm-cli/lib/workspaces/get-workspaces.js] [minimatch]: 0.055ms
MODULE_CJS 413480 [/home/h4ad/Projects/opensource/performance-analysis/npm-cli/lib/workspaces/get-workspaces.js] [@npmcli/package-json]: 0.044ms
MODULE_CJS 413480 [/home/h4ad/Projects/opensource/performance-analysis/npm-cli/lib/base-command.js] [./workspaces/get-workspaces.js]: 0.346ms
MODULE_CJS 413480 [/home/h4ad/Projects/opensource/performance-analysis/npm-cli/lib/base-command.js] [./utils/cmd-list]: 0.027ms
MODULE_CJS 413480 [/home/h4ad/Projects/opensource/performance-analysis/npm-cli/lib/base-command.js] [./utils/log-shim.js]: 0.012ms
MODULE_CJS 413480 [/home/h4ad/Projects/opensource/performance-analysis/npm-cli/lib/commands/run-script.js] [../base-command.js]: 0.638ms
MODULE_CJS 413480 [/home/h4ad/Projects/opensource/performance-analysis/npm-cli/lib/npm.js] [./commands/run-script.js]: 12.02ms
```

If we try to clean a little bit:

```bash
$ cat trace-node.txt | tr . , | sort -sbng -k5,16n -t ' ' | tr , . > formatted-timing.txt
```

The output was:

```
MODULE_TIMER 413856 [/home/h4ad/Projects/opensource/performance-analysis/npm-cli/node_modules/@npmcli/agent/lib/proxy.js] [socks-proxy-agent]: 13.046ms
MODULE_TIMER 413856 [/home/h4ad/Projects/opensource/performance-analysis/npm-cli/lib/npm.js] [./commands/run-script.js]: 13.548ms
MODULE_TIMER 413856 [/home/h4ad/Projects/opensource/performance-analysis/npm-cli/lib/cli-entry.js] [./npm.js]: 16.606ms
MODULE_TIMER 413856 [/home/h4ad/Projects/opensource/performance-analysis/npm-cli/node_modules/@npmcli/agent/lib/index.js] [./proxy.js]: 18.73ms
MODULE_TIMER 413856 [/home/h4ad/Projects/opensource/performance-analysis/npm-cli/node_modules/make-fetch-happen/lib/cache/entry.js] [cacache]: 19.421ms
MODULE_TIMER 413856 [/home/h4ad/Projects/opensource/performance-analysis/npm-cli/node_modules/make-fetch-happen/lib/remote.js] [@npmcli/agent]: 20.019ms
MODULE_TIMER 413856 [/home/h4ad/Projects/opensource/performance-analysis/npm-cli/node_modules/make-fetch-happen/lib/cache/entry.js] [../remote.js]: 21.397ms
MODULE_TIMER 413856 [/home/h4ad/Projects/opensource/performance-analysis/npm-cli/node_modules/make-fetch-happen/lib/cache/index.js] [./entry.js]: 42.496ms
MODULE_TIMER 413856 [/home/h4ad/Projects/opensource/performance-analysis/npm-cli/node_modules/make-fetch-happen/lib/fetch.js] [./cache/index.js]: 42.8ms
MODULE_TIMER 413856 [/home/h4ad/Projects/opensource/performance-analysis/npm-cli/node_modules/make-fetch-happen/lib/index.js] [./fetch.js]: 46.18ms
MODULE_TIMER 413856 [/home/h4ad/Projects/opensource/performance-analysis/npm-cli/node_modules/npm-registry-fetch/lib/index.js] [make-fetch-happen]: 47.044ms
MODULE_TIMER 413856 [/home/h4ad/Projects/opensource/performance-analysis/npm-cli/lib/utils/replace-info.js] [npm-registry-fetch]: 64.717ms
MODULE_TIMER 413856 [/home/h4ad/Projects/opensource/performance-analysis/npm-cli/lib/utils/error-message.js] [./replace-info.js]: 64.867ms
MODULE_TIMER 413856 [/home/h4ad/Projects/opensource/performance-analysis/npm-cli/lib/utils/exit-handler.js] [./error-message.js]: 73.463ms
MODULE_TIMER 413856 [/home/h4ad/Projects/opensource/performance-analysis/npm-cli/lib/cli-entry.js] [./utils/exit-handler.js]: 85.115ms
MODULE_TIMER 413856 [] [/home/h4ad/Projects/opensource/performance-analysis/npm-cli/bin/npm-cli.js]: 117.08ms
```

In this way, is very clear what is taking a lot of time to load `npm`, and I actually already opened a PR to fix one of the issues that appeared in this log (https://github.com/npm/cli/pull/7314)

---

@joyeecheung suggest to include also the support for `trace_events`, so I created the `debugWithTimer` to introduce support for both ways extract information, via log or via trace:

```bash
$ ./out/Release/node --trace-event-categories node.module_timer --trace-event-file-pattern 'trace-events.log' ~/Projects/opensource/performance-analysis/npm-cli/bin/npm-cli.js run echo --silent
```

This will generate a `trace-events.log` file that can be imported on `chrome://tracing` or [Perfetto UI](https://ui.perfetto.dev/) to generate the following graph:

![image](https://github.com/nodejs/node/assets/12551007/cab8fd0c-115c-4f64-8e3e-ce3145d18f24)

To be easier to read, I introduce two labels on the `debugWithTimer`, the first one will be for logging via `NODE_DEBUG` and the second one will be used at the `trace`.

/cc @nodejs/performance

GeoffreyBooth commented:
Could this perhaps get its own label in `NODE_DEBUG`? I feel like optimizing module load time is an infrequent use case for debugging modules and these logging lines must be very spammy.

joyeecheung commented on lib/internal/modules/cjs/loader.js:
```diff
@@ -966,10 +967,12 @@ function getExportsForCircularRequire(module) {
  * @param {boolean} isMain Whether the module is the main entry point
  */
 Module._load = function(request, parent, isMain) {
+  const start = performance.now();
```
`performance.now()` in itself leads to an overhead. I think we need to guard these behind a condition or a flag, or invent a specifier in `debuglog()` that leads to timing computation only when enabled.

Also as the test failure indicates this now load a bunch more perf-hooks modules unconditionally which slows down startup.

richardlau commented:
Conceptually this feels more like something for trace_events than NODE_DEBUG?

H4ad commented:
@richardlau Maybe I'm not familiar with trace_events but from what I see in the source code, I think it won't be as useful as having it as `NODE_DEBUG` since to be able to listen to the events we need to import `trace_events` and then `createTracing`.

The benefits of emitting these events through `NODE_DEBUG` is because it is very easy to listen to these events without any additional code, plus even importing `trace_events` will affect the events that will be emitted by this new `debug`, so I don't ' I don't think the implementation with `trace_events` will be as useful as this.

joyeecheung commented:
`createTracing()` is for user-land tracing to get hooked into our tracing infra. Pretty sure internally we can just use `internalBinding('trace_events').trace`, which is already what we load during startup.

joyeecheung commented on lib/internal/modules/cjs/loader.js:
```diff
@@ -359,10 +360,22 @@ function setModuleParent(value) {
 });
 
+let hrtimeBigIntTimingFn = () => {
+  hrtimeBigIntTimingFn = isDebugEnabled('timing_module_cjs') ? process.hrtime.bigint : () => 0;
```
I think it would make more sense to implement this in debuglog. Something like:

```js
const { debuglogWithTimer } = require('internal/util/debuglog');
const timer = debuglogWithTimer('timing_module_cjs');

// Say in `Module._load`, 'mark' should be request i.e. the module specifier/filename
timer.startTimer('mark');
// do some stuff
timer.logTimer('mark');
// do some stuff
timer.logTimer('mark');
// do some stuff
timer.logTimer('mark');
```

Then it'll be easy to reuse this in other places. And the `debuglogWithTimer` can emit traces via `internalBinding('trace_events').trace`.

H4ad replied:
I liked the idea but I didn't like the API, if we do it this way, we need to keep the start time in memory to be able to have the message `took 2ms`.

Or, we can just log the `hrtime` in every log and remove the necessity of `startTimer`:

```js
const mark = `[${ parent?.id }] [${ request }]`;

logTimer(mark, 'started');
// load the module
logTimer(mark, 'finished');
```

```
{debug_module} {pid}: HRTIME [{hrtime}] {mark} {message} 
```

In this way, what you did will be transformed as:

```
TIMING_MODULE_CJS 114493: HRTIME [1000] [/home/user/project/file.js] [path] started
TIMING_MODULE_CJS 114493: HRTIME [5000] [/home/user/project/file.js] [path] finished
```

H4ad replied:
I tried to create an implementation using `trace` and I don't think this is a good idea anymore because to be able to emit traces, we need to run the application like:

```
NODE_DEBUG=timing_module_cjs node --trace-categories node.timing_module index.js
```

And we can't enable the `trace` without enabling the `NODE_DEBUG`.

If we want to do that, we need to change how the `NODE_DEBUG` is enabled to make it work (and not set `noop`) when both `NODE_DEBUG` or `trace_events` is enabled.

Do you think that is worth implementing?

H4ad commented:
@joyeecheung I changed the implementation to match what we discussed, I liked this new way, it will print in the `stderr` if the `NODE_DEBUG=module_cjs` or it will trace if the category is `node. debuglog_{set}`, in this case, `node.debuglog_module_cjs`.

I still have some tests to fix but in general the implementation is good enough to get some reviews.

H4ad commented:
![image](https://github.com/nodejs/node/assets/12551007/5416618a-b9f3-4d86-9bf2-43d3cbad2691)

If we use the `chrome://tracing`, we can see what module is loading which module (I think we can improving this visualization).

H4ad commented:
I fixed the tests and I also changed a little bit the visualization for `trace`, I updated the PR to include the new graph and the new logging example.

joyeecheung commented on test/parallel/test-bootstrap-modules.js:
```diff
@@ -44,6 +44,7 @@ expected.beforePreExec = new Set([
   'NativeModule internal/assert',
   'NativeModule internal/util/inspect',
   'NativeModule internal/util/debuglog',
+  'NativeModule internal/util/trace_timer',
```
This should only be loaded when the timer is enabled.

H4ad replied:
`trace_timer` is also imported by `Console`, so that's why is loaded every time.

Do you want to lazy load even for `Console`?

H4ad replied:
Done!

RafaelGSS commented on lib/internal/modules/cjs/loader.js:
```diff
@@ -966,10 +973,15 @@ function getExportsForCircularRequire(module) {
+  const logLabel = `[${parent?.id || ''}] [${request}]`;
+  const traceLabel = `require('${request}')`;
+
+  startTimer(logLabel, traceLabel);
```
Shouldn't it check if trace events are enabled first? It seems to affect (in terms of performance) all users in general, not just NODE_DEBUG

H4ad replied:
I do the check at https://github.com/nodejs/node/pull/52213/files#diff-2b09f603b3132191141319f8bfa6d5af325d40d0827a9fc10f1757af9fe21cafR218-R220

Since there are two conditions to enable (log and trace), I need to check it later.

joyeecheung replied:
The typical pattern of `debuglog` and friend is, when NODE_DEBUG is enabled, on first invocation the callback of `debugWithTimer` would replace `startTimer` with a function that does something. Otherwise it's replaced with noops which V8 can inline away.

H4ad replied:
Since we are integrating with `trace_events`, there's no way to avoid perf penalty here and without having a listener to the `trace_events` (I don't think that exist), we can't replace with `noop` functions.

H4ad replied:
Ok, as soon I have time I will rewrite this to change to noop in the initialization, then I will create a PR to fix this issue based on https://github.com/nodejs/node/pull/48142

joyeecheung replied:
Yes I think a use case like this would justify something like https://github.com/nodejs/node/pull/48142 - doing the trace enabling check is going to lead to a bigger perf hit on module loading than it does on HTTP (it didn't actually look like that much code to me, BTW, it's mostly just boilerplate binding code, but maybe I am biased because I am too used to boilerplate binding code).

RafaelGSS reviewed (commented):
Can you update the PR to include only the changes this PR is about? I can see some linting and type changes.

joyeecheung commented on lib/internal/util/debuglog.js:
```diff
@@ -115,7 +147,159 @@ function debuglog(set, cb) {
+    init();
+
+    if (cb)
+      cb(internalStartTimer, internalEndTimer, internalLogTimer);
```
It should do something even cheaper - detect whether the environment variables are enabled, if not, do `cb(noop, noop, noop)` - and also once the `cb` is invoked, don't invoke it again, the same should also be done for the two other methods. This is what we do for `debuglog` (`debuglogImpl()` returns the `noop` constants which gets re-assigned as the variable in the caller via the callback).

H4ad replied:
I can't do that, `trace_events` can be enabled at runtime, so there's no way to avoid using the original implementation here and I always need to check if the category is enabled.

joyeecheung commented on lib/internal/util/trace_timer.js:
```diff
@@ -0,0 +1,185 @@
+  StringPrototypePadStart,
+} = primordials;
+
+const { trace } = internalBinding('trace_events');
```
Note that adding an unsnapshotted new file leads to a performance hit on first load. It doesn't seem necessary to split a new file for these, they can probably be just moved into `lib/internal/util/debuglog.js`

H4ad replied:
I split in another file to be able to reuse this functionality on `Console`, and `Console` don't need to import `debuglog`.

H4ad replied:
In the end, I moved to the `debuglog` since I introduced `flags` to the `timer` functions to skip calling C++.

H4ad commented:
I moved the new code entirely to `debuglog` as suggested by Joyee, I added more tests to ensure both `trace-events` and `NODE_DEBUG` are working correctly, and I also checked only startup time the presence of `trace-categories` flags and then replaced the functions to `noop`.

I leave a TODO for myself to optimize the `isTraceCategoryEnabled` to let this feature to be enabled at runtime.

joyeecheung commented on lib/internal/console/constructor.js:
```diff
@@ -62,18 +57,25 @@ const {
+    return _debugLog;
+  }
+
+  _debugLog = require('internal/util/debuglog');
```
Since it's in `internal/util/debuglog` now, it's fine to not lazy load this.

joyeecheung commented on lib/internal/modules/cjs/loader.js:
```diff
@@ -1021,6 +1039,8 @@ Module._load = function(request, parent, isMain) {
     }
```
We might also want to end the timer in the throw statement above? Not sure what it would look like in the trace if someone try-catch it...

H4ad replied:
Good catch

joyeecheung commented on lib/internal/util/debuglog.js:
```diff
@@ -115,7 +126,367 @@ function debuglog(set, cb) {
+      );
+  }
+
+  const kTraceCategory = buildCategory(StringPrototypeToLowerCase(set));
```
Any reason not to just inline this?

H4ad replied:
No, I think in the beginning I was planning to use this method in other places but I think is fine to inline.

joyeecheung commented on lib/internal/util/debuglog.js:
```diff
@@ -115,7 +126,367 @@ function debuglog(set, cb) {
+    init();
+
+    if (debugLogCategoryEnabled || traceCategoryEnabled)
+      return internalStartTimer(logLabel, traceLabel);
```
I don't think we need to return the functions invocation results here? It's better to not return them if the return values aren't needed, lest it gives the impression that the return value is important somehow.

joyeecheung reviewed (approved):
Great work! LGTM with some nits

H4ad commented:
@joyeecheung Should we support ending the trace even if the `_load` throws an unexpected error?

I was creating a trace for `pnpm` and for some reason, the trace was not ending, I changed the `_load` code to have a huge `try/finally` and then the `trace` was fixed.

Is this a scenario we want to support or we can ignore it?

joyeecheung commented:
> even if the _load throws an unexpected error?

Where is that error coming from? I think most error paths are ending the timers here already? (especially the try-catch around `module.load`)?

H4ad commented:
From what I debug, the error is thrown when the module is not found.

Also, I read the source code of the functions that are called and there are other cases where it throws an error without a try-catch, like inside `getExportsForCircularRequire`, those cases will cause the trace to start but not end.

I added a huge try-catch to handle those cases to always end a trace.

joyeecheung commented:
Hmm I see, the inner functions can throw too.

Can we minimize the changes by adding an internal helper in cjs/loader.js that wraps Module._load with a try-finally block? Then change the internals that call Module._load to call this helper instead. That was also what we did with wrapSafe etc.

joyeecheung commented:
Actually putting try-catch around `Module._load` might have the additional benefit that...if users monkey-patch it (😅) we can also measure the time took in the monkey-patched version (e.g. with [Yarn pnp](https://github.com/yarnpkg/berry/blob/c64644ae88ba42090ab9b87da0b09601d4218b0c/packages/yarnpkg-pnp/sources/loader/applyPatch.ts#L61))

H4ad commented:
I added the `wrapModuleLoad` to implement the logic of tracing & recursion tracking.

But I also noticed a lot of places where `_load` is called directly, should we rewrite those places to call `wrapModuleLoad`?

Those places don't call `requireDepth`, so I don't think is planned, or if we change those places to call `wrapModuleLoad`, it will be a breaking change since is changing the `requireDepth`.

joyeecheung commented:
> But I also noticed a lot of places where _load is called directly, should we rewrite those places to call wrapModuleLoad?

Yes.

> Those places don't call requireDepth, so I don't think is planned, or if we change those places to call wrapModuleLoad, it will be a breaking change since is changing the requireDepth.

AFAICT those places aren't related to `require()` anyway (e.g. it's import cjs), so it's okay that they are not doing `requireDepth`. Probably out of the scope of this PR to try to fix them.

joyeecheung commented on lib/internal/modules/cjs/loader.js:
```diff
@@ -209,12 +238,7 @@ function internalRequire(module, id) {
-  } finally {
-    requireDepth--;
-  }
+  return wrapModuleLoad(id, module, /** isMain */ false);
```
This should still keep the original try-finally with just the  `Module._load` replaced with `wrapModuleLoad()`, and only do the requireDepth computation in `internalRequire` and `Module.prototype.require`, so that other code that does `Module._load` can also just replace it with `wrapModuleLoad()`

H4ad commented:
Sorry for taking so long to fix the issues, I was taking a break from OSS.

I fixed all the changes on the snapshot tests, since we introduced a new function in the call stack, I think the `semver-minor` already covers the "potential breaking change" of this new function.

This PR got bigger than I expected but I think is a worthy feature to have, once we merge it, I will start working on the support for ESM and dynamic enable of trace feature.

mcollina reviewed (approved):
rlgtm

aduh95 commented:
This needs a rebase

H4ad commented:
@joyeecheung The current CI is failing because the `trace` is emitting every string without escaping, so something like this:

```js
console.time('C:\\wddd');

setTimeout(() => console.timeEnd('C:\\wddd'), 200);
```

Will emit an invalid JSON even on v22 because the output string will be:

```json
"name":"time::C:\wddd",
```

Probably `.replace('\\', '\\\\')` on the `trace` label should be enough to fix this issue, and since we merge the implementation of `console.time` with `debugWithTimer`, this will fix both cases.

Any hints on how to proceed?

H4ad commented:
I added escaping for `\\` and also a test to ensure it's working for `console.time`.

I also tried to fix all issues on Windows, so now it _should_ pass all the tests (hopefully)

joyeecheung commented on lib/internal/modules/cjs/loader.js:
```diff
@@ -189,6 +197,25 @@ let requireDepth = 0;
 let statCache = null;
 
+/**
+ * Internal method to be able to add recursion & tracing capabilities
```
Why is recursion mentioned here? AFAICT it is unrelated?

H4ad replied:
I removed the `requireDepth++`, so this comment is outdated, I will remove it.

joyeecheung commented on lib/internal/modules/cjs/loader.js:
```diff
@@ -181,6 +182,13 @@ const {
 
+const { debuglog, debugWithTimer } = require('internal/util/debuglog');
+
+let { startTimer, endTimer } = debugWithTimer('module_cjs', (start, end) => {
```
Should this be:

```suggestion
let { startTimer, endTimer } = debugWithTimer('module_timer', (start, end) => {
```

Instead? I feel that if we are to add something similar to ESM, there's not too much point to distinguish the two env vars. In the traces users can distinguish how the module get loaded with the log labels anyway.

H4ad replied:
Hm.. makes sense to me, I initially thought the algorithm would accept if we pass `node.debuglog_module` to enable both `_cjs` and `_esm`, but I think it's not the case.

joyeecheung commented on lib/internal/util/debuglog.js:
```diff
@@ -115,7 +126,371 @@ function debuglog(set, cb) {
+      );
+  }
+
+  const kTraceCategory = `node,node.debuglog_${StringPrototypeToLowerCase(set)}`;
```
Do we really need the `debuglog_` prefix though? It seems fine to just use the label directly as category.

H4ad replied:
If we plan to use this helper in other places that already have tracing capabilities, maybe having this directly to `node.${set}` will be better since the user only needs to enable one category.

Initially, I wanted to have this completely separated to avoid collisions, but if we consider the above use-case, could make sense to remove the prefix.

joyeecheung reviewed (commented):
Codewise LGTM, some questions about the env var/category names

H4ad commented:
I got some errors from CI, one of them is:

```
not ok 2208 parallel/test-module-print-timing
18:02:18   ---
18:02:18   duration_ms: 340.76000
18:02:18   severity: fail
18:02:18   exitcode: 1
18:02:18   stack: |-
18:02:18     TAP version 13
18:02:18     # Subtest: should print the timing information for cjs
18:02:18     not ok 1 - should print the timing information for cjs
18:02:18       ---
18:02:18       duration_ms: 11.76897
18:02:18       location: '/home/iojs/build/workspace/node-test-commit-linux-containered/test/parallel/test-module-print-timing.mjs:10:1'
18:02:18       failureType: 'testCodeFailure'
18:02:18       error: |-
18:02:18         Expected values to be strictly equal:
18:02:18         
18:02:18         127 !== 0
18:02:18         
18:02:18       code: 'ERR_ASSERTION'
18:02:18       name: 'AssertionError'
18:02:18       expected: 0
18:02:18       actual: 127
18:02:18       operator: 'strictEqual'
18:02:18       stack: |-
18:02:18         TestContext.<anonymous> (file:///home/iojs/build/workspace/node-test-commit-linux-containered/test/parallel/test-module-print-timing.mjs:17:10)
18:02:18         async Test.run (node:internal/test_runner/test:842:9)
18:02:18         async Test.processPendingSubtests (node:internal/test_runner/test:550:7)
```

> Link: https://ci.nodejs.org/job/node-test-commit-linux-containered/nodes=ubuntu2204_sharedlibs_openssl30_x64/43597/consoleFull

Which is something like: `usually indicates that the command or binary being executed within a container could not be found` which does not make sense since I'm using `execPath` and `-e`.

The other one is on Windows, the build is just failing and I didn't find the reason: https://ci.nodejs.org/job/node-compile-windows-debug/nodes=win-vs2022/22042/consoleFull

@nodejs/build any hints?

joyeecheung commented on test/parallel/test-module-print-timing.mjs:
```diff
@@ -0,0 +1,91 @@
+    },
+  });
+
+  assert.strictEqual(result.code, 0);
```
You might want to log the stdout and stderr instead of going straight to asserting the exit code for getting more information out of the CI, or use a utility like `spawnSyncAndAssert` in `common/child_process` which is designed to log more things in the CI on mismatch.

H4ad replied:
I'm getting the following error:

```
 /home/iojs/build/workspace/node-test-commit-linux-containered/out/Release/node: error while loading shared libraries: libcrypto.so.81.3: cannot open shared object file: No such file or directory
22:44:32     
22:44:32     [process 1608929]: --- stdout ---
22:44:32     
22:44:32     [process 1608929]: status = 127, signal = null
22:44:32     [process 1608953]: --- stderr ---
22:44:32     /home/iojs/build/workspace/node-test-commit-linux-containered/out/Release/node: error while loading shared libraries: libcrypto.so.81.3: cannot open shared object file: No such file or directory
22:44:32     
22:44:32     [process 1608953]: --- stdout ---
```

I tried to change the `execPath` but it didn't seems to change the error :/

joyeecheung replied:
Does the tracing stuff rely on crypto somehow? There might be something wrong with the trace events in shared OpenSSL mode. You can try skipping tracing in those builds.

richardlau replied:
Don't replace env -- you'll lose `LD_LIBRARY_PATH` which will prevent the OpenSSL libraries (or any of the shared libs for [node-test-commit-linux-containered](https://ci.nodejs.org/job/node-test-commit-linux-containered/)) from being found. Extend it instead (add/removing as necessary).

H4ad commented:
@joyeecheung We are closer, just one (re)approval :)

joyeecheung reviewed (approved):
Nice job! LGTM

rluvaton commented:
@H4ad Is it possible to backport this to node 20?

H4ad commented:
I don't know, usually we need someone from the release team to select this PR first to see if there is any conflict, only if we have conflicts we need a manual backport.

joyeecheung commented:
Tentatively backported in https://github.com/nodejs/node/pull/56927 - backporting this reduces the conflicts, though it seems associated with some other later commits, some included regressions, so only tentatively backporting it together with the associated commits that I found, which seemed safe enough.

joyeecheung commented:
Backing it out from https://github.com/nodejs/node/pull/56927 to avoid introducing surface for regressions.
